lib/db.js

```diff
@@ -15,181 +15,374 @@
+		this.system.on('db_save', () => {
+			// Not wired ... interval has control
+			//this.save()
+		})
+
+		this.load()
 
-		var changed_after_load = false
 		// db defaults
-		if (self.db.userconfig === undefined) {
-			self.db.userconfig = {}
-			changed_after_load = true
+		if (this.store.userconfig === undefined) {
+			this.store.userconfig = {}
+			this.setDirty()
 		}
 
 		// is page up 1->2 or 2->1?
-		if (self.db.userconfig.page_direction_flipped === undefined) {
-			self.db.userconfig.page_direction_flipped = false
-			changed_after_load = true
-		}
-
-		system.emit('db_loaded', self.db)
-
-		if (changed_after_load === true) {
-			debug('config changed by default values after load, saving.')
-			system.emit('db_save')
-		}
-	} catch (err) {
-		if (err.code == 'ENOENT') {
-			debug('readFile(db)', 'Couldnt read db, loading {}')
-			system.emit('db_loaded', {})
-		} else {
-			throw err
+		if (this.store.userconfig.page_direction_flipped === undefined) {
+			this.store.userconfig.page_direction_flipped = false
+			this.setDirty()
 		}
 	}
 
-	system.on('db_all', function (cb) {
-		debug('db_all(): returning all database values')
-		if (typeof cb == 'function') {
-			cb(self.db)
-		}
-	})
-
-	system.on('db_set', function (key, value) {
-		debug('db_set(' + key + ', ' + value + ')')
-		self.db[key] = value
-	})
-
-	system.on('db_del', function (key) {
-		debug('db_del(' + key + ')')
-		delete self.db[key]
-	})
-
-	system.on('db_set_multiple', function (keyvalueobj) {
-		debug('db_set_multiple:')
-		for (var key in keyvalueobj) {
-			debug('db_set(' + key + ',' + keyvalueobj[key] + ')')
-			self.db[key] = keyvalueobj[key]
-		}
-	})
-
-	system.on('db_get', function (key, cb) {
-		debug('db_get(' + key + ')')
-		cb(self.db[key])
-	})
-	system.on('db_get_multiple', function (keys, cb) {
-		if (typeof keys != 'object' || typeof keys.length == 'undefined') {
-			throw new Error('keys is not an array')
+	/**
+	 * Delete a key/value pair
+	 * @param {string} key - the key to be delete
+	 * @access public
+	 */
+	deleteKey(key) {
+		this.debug(`${this.name}_del (${key})`)
+		if (key !== undefined) {
+			delete this.store[key]
+			this.setDirty()
 		}
-		cb(
-			keys.map(function (key) {
-				return self.db[key]
-			})
-		)
-	})
+	}
 
-	async function doSave() {
-		const jsonStr = JSON.stringify(self.db)
+	/**
+	 * Save the database to file making a `FILE.bak` version then moving it into place
+	 * @async
+	 * @param {boolean} [withBackup = true] - can be set to <code>false</code> if the current file should not be moved to `FILE.bak`
+	 * @access protected
+	 */
+	async doSave(withBackup) {
+		const jsonSave = JSON.stringify(this.store)
+		this.dirty = false
+		this.lastsave = Date.now()
 
-		try {
-			await fs.copy(pathDb, pathDbBak)
-		} catch (err) {
-			debug('db_save', 'Error making backup of config: ' + err)
+		if (withBackup === true && fs.existsSync(this.cfgFile) && fs.readFileSync(this.cfgFile, 'utf8').trim().length > 0) {
```

Comment:
this has been updated, and since I read the current DB in to check for an empty file I decided to attempt a parse so we don't backup a corrupted file.  thoughts?

Reply:
I like it :)